### PR TITLE
Health check shield API

### PR DESCRIPTION
**Pull request by ezsilmar**
Draft implementation of the API which temporarily prevents the task from being killed when it's unhealthy.
This is required to support troubleshooting scenarios when the app is "legally" unhealthy, for example when taking a memory dump.

API:
- GET /v2/shield: show current shields
- PUT /v2/shield/<task_id>?ttl=<ttl>: add the shield
- DELETE /v2/shield/<task_id>: remove the shield

TODO:
- do we need to persist shield state ?
- authorization
- ttl tests

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/ShieldResource.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import mesosphere.marathon.raml.Raml
+import mesosphere.marathon.raml.HealthConversion._
+
+import scala.async.Async._
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+@Path("v2/shield")
+@Produces(Array(MediaType.APPLICATION_JSON))
+class HealthCheckShieldResource @Inject() (
```
Comment (Lqp1) [inline src/main/scala/mesosphere/marathon/api/v2/ShieldResource.scala]: It's cosmetic but should not the file be named after class name?

File: src/main/scala/mesosphere/marathon/core/health/HealthCheckShield.scala [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package mesosphere.marathon
+package core.health
+
+import mesosphere.marathon.state.Timestamp
+import mesosphere.marathon.core.task.Task
+
+case class HealthCheckShield(taskId: Task.Id, ttl: Timestamp) {
```
Comment (Lqp1) [inline src/main/scala/mesosphere/marathon/core/health/HealthCheckShield.scala]: Another cosmectic point: to me ttl refers to a duration, but at this point, you changed it to be now+ttl, so it's not a TTL anymore. Completely optional, but maybe we should find a better suited name for this field. Eg: expiryTime, endTime, ...

File: src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckShieldManager.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import mesosphere.marathon.core.health.HealthCheckShield
+
+import scala.concurrent.duration._
+import scala.collection.concurrent.TrieMap
+
+class HealthCheckShieldManager extends Object with StrictLogging {
+
+  private[this] val shields = TrieMap.empty[Task.Id, HealthCheckShield]
+
+  def isShielded(taskId: Task.Id): Boolean = {
```
Comment (Lqp1) [inline src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckShieldManager.scala]: FAR: I think (i'm not 100% sure) most shield info will never get removed because marathon would call isShielded only for unhealthy instances. So for example, in most cases (where app becomes healthy again once heap dump is finished) the TLL is expired but it will never be removed.
Memory usage is quite low so it's acceptable, but there is an area of improvement here (like a quick shield check on HealthResult message) :)

Also, there are leader changes (= marathon leader restarts) quite often, so it should not be a big issue.

**Review by Lqp1 — COMMENTED** [review]
Looks good overall :)
To answer your question in commit message, I think we should persist the shield information; there are plenty of leader changes, and abdicating marathon leader fully restarts in that case. If it's expected to protect marathon apps for long heap dump procedures, it should then be persisted, otherwise app would be killed by new leader.

File: src/main/scala/mesosphere/marathon/api/v2/ShieldResource.scala (hunk 1 above)
Comment (ezsilmar) [reply src/main/scala/mesosphere/marathon/api/v2/ShieldResource.scala]: Done!

File: src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckShieldManager.scala (hunk 3 above)
Comment (ezsilmar) [reply src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckShieldManager.scala]: Introduced periodical "purging" of the expired shields. 
It looks harder to react on tasks health events + introduces more coupling with existing marathon code.

File: src/main/scala/mesosphere/marathon/api/v2/HealthCheckShieldResource.scala [hunk 4]
```diff
@@ -29,8 +29,8 @@ class HealthCheckShieldResource @Inject() (
   @GET
   def index(
     @Context req: HttpServletRequest, @Suspended asyncResponse: AsyncResponse): Unit = sendResponse(asyncResponse) {
-    async{
-      val shields = healthCheckManager.listShields()
+    async {
+      val shields = await (healthCheckManager.listShields())
```
Comment (Lqp1) [inline src/main/scala/mesosphere/marathon/api/v2/HealthCheckShieldResource.scala]: Maybe we should check somehow the return of the future in each endpoint and return adapted status code? I think before it was caught as an exception but now I'm not sure with the future.

**Review by Lqp1 — COMMENTED** [review]
I'm not sure about the zk part but otherwise looks good, thanks!

File: src/main/scala/mesosphere/marathon/api/v2/HealthCheckShieldResource.scala (hunk 4 above)
Comment (ezsilmar) [reply src/main/scala/mesosphere/marathon/api/v2/HealthCheckShieldResource.scala]: I checked that if the exception happens it propagates here and we return 503

Comment (Lqp1) [reply src/main/scala/mesosphere/marathon/api/v2/HealthCheckShieldResource.scala]: Thanks!

File: docs/docs/rest-api/public/api/v2/shield.raml [hunk 5]
```diff
@@ -49,9 +49,15 @@ description: |
       200:
         description: |
           The shield was successfully created or updated.
-      503:
+      400:
         description: |
-          Health check shield API is disabled.
```
Comment (Lqp1) [inline docs/docs/rest-api/public/api/v2/shield.raml]: Just a question: why did you remove 503 return code? I would expect it to still be returned if the feature is disabled?

Comment (Lqp1) [reply docs/docs/rest-api/public/api/v2/shield.raml]: Ah ok nvm, you merged the forbidden because of feature is disabled with forbidden because user does not have the rights to shield.

**Review by Lqp1 — APPROVED** [review]

**Comment by komuta** [comment]
manually squash-merged
